flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalogBase.java

```diff
@@ -407,4 +443,201 @@ private Table getHiveTable(ObjectPath tablePath) throws TableNotExistException {
+		Table hiveTable = getHiveTable(tablePath);
+
+		ensurePartitionedTable(tablePath, hiveTable);
+
+		try {
+			// partition spec can be partial
+			List<String> partialVals = MetaStoreUtils.getPvals(hiveTable.getPartitionKeys(), partitionSpec.getPartitionSpec());
+			return client.listPartitionNames(tablePath.getDatabaseName(), tablePath.getObjectName(), partialVals,
+				(short) -1).stream().map(HiveCatalogBase::createPartitionSpec).collect(Collectors.toList());
+		} catch (TException e) {
+			throw new CatalogException(
+				String.format("Failed to list partitions of table %s", tablePath), e);
+		}
+	}
+
+	@Override
+	public CatalogPartition getPartition(ObjectPath tablePath, CatalogPartitionSpec partitionSpec)
+		throws PartitionNotExistException, CatalogException {
+		try {
+			Table hiveTable = getHiveTable(tablePath);
+			ensurePartitionedTable(tablePath, hiveTable);
+			Partition hivePartition = client.getPartition(tablePath.getDatabaseName(), tablePath.getObjectName(),
+				getFullPartitionValues(tablePath, partitionSpec, getFieldNames(hiveTable.getPartitionKeys())));
+			return createCatalogPartition(hivePartition);
+		} catch (NoSuchObjectException e) {
+			throw new PartitionNotExistException(catalogName, tablePath, partitionSpec);
+		} catch (TException e) {
+			throw new CatalogException(
+				String.format("Failed to get partition %s of table %s", partitionSpec, tablePath), e);
+		} catch (TableNotPartitionedException | TableNotExistException | PartitionSpecInvalidException e) {
+			throw new PartitionNotExistException(catalogName, tablePath, partitionSpec, e);
+		}
+	}
+
+	@Override
+	public void alterPartition(ObjectPath tablePath, CatalogPartitionSpec partitionSpec, CatalogPartition newPartition, boolean ignoreIfNotExists)
+		throws PartitionNotExistException, CatalogException {
+		// Explicitly check if the partition exists or not
+		// because alter_partition() doesn't throw NoSuchObjectException like dropPartition() when the target doesn't exist
+		if (partitionExists(tablePath, partitionSpec)) {
+			try {
+				Table hiveTable = getHiveTable(tablePath);
+				ensurePartitionedTable(tablePath, hiveTable);
+				client.alter_partition(
+					tablePath.getDatabaseName(),
+					tablePath.getObjectName(),
+					createHivePartition(hiveTable, partitionSpec, newPartition)
+				);
+			} catch (TException e) {
+				throw new CatalogException(
+					String.format("Failed to alter existing partition with new partition %s of table %s",
+						partitionSpec, tablePath), e);
+			} catch (TableNotPartitionedException | TableNotExistException | PartitionSpecInvalidException e) {
+				throw new PartitionNotExistException(catalogName, tablePath, partitionSpec, e);
+			}
+		} else if (!ignoreIfNotExists) {
+			throw new PartitionNotExistException(catalogName, tablePath, partitionSpec);
+		}
+	}
+
+	private void ensurePartitionedTable(ObjectPath tablePath, Table hiveTable) throws TableNotPartitionedException {
```

Comment:
1. can this be static?
2. Can this be combined with another method of the same name?

Reply:
It can be static as long as we pass all the needed (non-static) fields to it as parameters. The other method is in `GenericInMemoryCatalog`. Shall we have a class for different catalogs to share util methods?

Reply:
I'd say making a method static as long as it doesn't need to access instance members. For utils, I don't see why not if there is a case for sharing.